ml-proto/spec/decode.ml

```diff
@@ -0,0 +1,602 @@
+
+
+(* Sections *)
+
+let trace s name =
+  print_endline
+    (name ^ " @ " ^ string_of_int (pos s) ^ " = " ^ string_of_byte (read s))
+
+let id s =
+  match string s with
+  | "type" -> `TypeSection
+  | "import" -> `ImportSection
+  | "function" -> `FuncSection
+  | "table" -> `TableSection
+  | "memory" -> `MemorySection
+  | "export" -> `ExportSection
+  | "start" -> `StartSection
+  | "code" -> `CodeSection
+  | "data" -> `DataSection
+  | "end" -> `EndSection
+  | _ -> `UnknownSection
+
+let section tag f default s =
+  let start_pos = pos s in
+  let size = vu s in
+  let id_pos = pos s in
+  if id s <> tag then (rewind start_pos s; default) else
+  let s' = substream s (id_pos + size) in
+  let x = f s' in
+  require (eos s') s' (pos s') "junk at end of section";
+  x
+
+
+(* Type section *)
+
+let type_section s =
+  section `TypeSection (vec func_type) [] s
+
+
+(* Import section *)
+
+let import s =
+  let itype = at var s in
+  let module_name = string s in
+  let func_name = string s in
+  {itype; module_name; func_name}
+
+let import_section s =
+  section `ImportSection (vec (at import)) [] s
+
+
+(* Function section *)
+
+let func_section s =
+  section `FuncSection (vec (at var)) [] s
+
+
+(* Table section *)
+
+let table_section s =
+  section `TableSection (vec (at var)) [] s
+
+
+(* Memory section *)
+
+let memory s =
+  let min = vu64 s in
+  let max = vu64 s in
+  let _ = bool s in (*TODO: pending change*)
+  {min; max; segments = []}
+
+let memory_section s =
+  section `MemorySection (opt (at memory) true) None s
+
+
+(* Export section *)
+
+let export s =
+  let x = at var s in
+  let name = string s in
+  {name; kind = `Func x} (*TODO: pending resolution*)
+
+let export_section s =
+  section `ExportSection (vec (at export)) [] s
+
+
+(* Start section *)
+
+let start_section s =
+  section `StartSection (opt (at var) true) None s
+
+
+(* Code section *)
+
+let local s =
+  let n = vu s in
+  let t = value_type s in
+  Lib.List.make n t
+
+let code s =
+  let locals = List.flatten (vec local s) in
+  let size = vu s in
+  let body = expr_block (substream s (pos s + size)) in
+  {locals; body; ftype = Source.((-1) @@ Source.no_region)}
+
+let code_section s =
+  section `CodeSection (vec (at code)) [] s
+
+
+(* Data section *)
+
+let segment s =
+  let addr = vu64 s in
+  let data = string s in
+  {Memory.addr; data}
+
+let data_section s =
+  section `DataSection (vec (at segment)) [] s
+
+
+(* End section *)
+
+let end_section s =
+  section `EndSection ignore () s
+
+
+(* Unknown section *)
+
+let unknown_section s =
+  section `UnknownSection (fun s -> skip (len s - pos s) s; true) false s
+
+
+(* Modules *)
+
+let rec iterate f s = if f s then iterate f s
+
+let module_ s =
+  let magic = u32 s in
+  require (magic = 0x6d736100l) s 0 "magic header not detected";
+  let version = u32 s in
+  require (version = 0x0bl) s 4 "unknown binary version";
+  iterate unknown_section s;
+  let types = type_section s in
```

Comment:
See, it wouldn't be so terrible to pass in `types` as another arg to `code_section` ;)